src/lang/frontend_ast_lower.rs

```diff
@@ -0,0 +1,798 @@
+            match arg {
+                StatementArg::Identifier(id) => {
+                    if !seen.contains(&id.name) {
+                        seen.insert(id.name.clone());
+                        names.push(id.name.clone());
+                    }
+                }
+                StatementArg::AnchoredKey(ak) => {
+                    if !seen.contains(&ak.root.name) {
+                        seen.insert(ak.root.name.clone());
+                        names.push(ak.root.name.clone());
+                    }
+                }
+                StatementArg::Literal(_) => {}
+            }
+        }
+    }
+
+    fn extract_and_split_predicates(
+        &self,
+    ) -> Result<(Vec<CustomPredicateDef>, usize), LoweringError> {
+        let doc = self.validated.document();
+        let predicates: Vec<CustomPredicateDef> = doc
+            .items
+            .iter()
+            .filter_map(|item| match item {
+                DocumentItem::CustomPredicateDef(pred) => Some(pred.clone()),
+                _ => None,
+            })
+            .collect();
+
+        let original_count = predicates.len();
+
+        // Apply splitting to each predicate as needed
+        let mut split_predicates = Vec::new();
+        for pred in predicates {
+            let chain = frontend_ast_split::split_predicate_if_needed(pred, self.params)?;
+            split_predicates.extend(chain);
+        }
+
+        Ok((split_predicates, original_count))
+    }
+
+    fn lower_custom_predicate(
+        &self,
+        pred_def: &CustomPredicateDef,
+        cpb_builder: &mut CustomPredicateBatchBuilder,
+    ) -> Result<(), LoweringError> {
+        let name = pred_def.name.name.clone();
+
+        // Note: Constraint checking is handled by the splitting phase
+        // Predicates passed here should already be within limits
+
+        // Collect public and private argument names
+        let mut public_arg_names = Vec::new();
+        let mut private_arg_names = Vec::new();
+
+        for arg in &pred_def.args.public_args {
+            public_arg_names.push(arg.name.clone());
+        }
+
+        if let Some(private_args) = &pred_def.args.private_args {
+            for arg in private_args {
+                private_arg_names.push(arg.name.clone());
+            }
+        }
+
+        // Lower statements to builders
+        let mut statement_builders = Vec::new();
+        for stmt in &pred_def.statements {
+            let stmt_builder = self.lower_statement_to_builder(stmt)?;
+            statement_builders.push(stmt_builder);
+        }
+
+        // Convert to &str slices for builder API
+        let public_args_str: Vec<&str> = public_arg_names.iter().map(|s| s.as_str()).collect();
+        let private_args_str: Vec<&str> = private_arg_names.iter().map(|s| s.as_str()).collect();
+
+        // Add predicate to batch using builder
+        let conjunction = pred_def.conjunction_type == ConjunctionType::And;
+
+        if conjunction {
```

Comment:
You could use `cpb_builder.predicate(name, conjunction, ...)` to avoid this `if/else`.